pallets/treasury-buyout-extension/src/lib.rs

```diff
@@ -0,0 +1,543 @@
+#![deny(warnings)]
+#![cfg_attr(test, feature(proc_macro_hygiene))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[cfg(test)]
+mod mock;
+
+pub mod default_weights;
+
+#[cfg(test)]
+mod tests;
+
+mod types;
+
+use crate::{
+	default_weights::WeightInfo,
+	types::{AccountIdOf, Amount, BalanceOf, CurrencyIdOf},
+};
+use codec::{Decode, Encode};
+use frame_support::{
+	dispatch::{DispatchError, DispatchResult},
+	ensure,
+	sp_runtime::SaturatedConversion,
+	traits::{Get, IsSubType},
+};
+use orml_traits::MultiCurrency;
+pub use pallet::*;
+use sp_arithmetic::{
+	per_things::Rounding,
+	traits::{CheckedAdd, Saturating},
+};
+use sp_runtime::{
+	traits::{DispatchInfoOf, One, SignedExtension, Zero},
+	transaction_validity::{
+		InvalidTransaction, TransactionValidity, TransactionValidityError, ValidTransaction,
+	},
+	ArithmeticError, FixedPointNumber, FixedU128,
+};
+use sp_std::{fmt::Debug, marker::PhantomData};
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::{pallet_prelude::*, sp_runtime::Permill};
+	use frame_system::{ensure_root, ensure_signed, pallet_prelude::*};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + orml_currencies::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Used for currency-related operations
+		type Currency: MultiCurrency<AccountIdOf<Self>, CurrencyId = CurrencyIdOf<Self>>;
+
+		/// Used for getting the treasury account
+		#[pallet::constant]
+		type TreasuryAccount: Get<AccountIdOf<Self>>;
+
+		/// Buyout period in blocks in which a caller can buyout up to the amount limit stored in `BuyoutLimit`
+		/// When attempting to buyout after this period, the buyout limit is reset for the caller
+		#[pallet::constant]
+		type BuyoutPeriod: Get<u32>;
+
+		/// Fee from the native asset buyouts
+		#[pallet::constant]
+		type SellFee: Get<Permill>;
+
+		/// Type that allows for checking if currency type is ownable by users
+		type AllowedCurrencyChecker: AllowedCurrencyChecker<CurrencyIdOf<Self>>;
```

Comment:
I think that is a great idea 👍🏼. Since we may open multiple new HRMP channels in the future, I think that in the long run we are going to save quite some time and effort by doing a root transaction instead of updating the runtime everytime we want to allow/deny an asset. Even if we don't change the allowed assets too often, in my opinion it's still worth storing the supported currencies.

Reply:
Added the changes for storing allowed currencies @ebma @gianfra-t